### PR TITLE
sql: support PREPARE, EXECUTE

### PR DESCRIPTION
**Pull request by mjibson**
### Motivation

  * This PR adds a known-desirable feature. #3383

### Description

PREPARE is analogous to pgwire Parse, EXECUTE is pgwire Bind+Execute, and DEALLOCATE is pgwire Close. These are SQL-level control of session state, similar to DECLARE and FETCH controlling session portals.

### Checklist

- [x] This PR has adequate test coverage.
- [ ] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

File: src/sql/src/plan/query.rs [hunk 1]
```diff
@@ -913,6 +913,59 @@ pub fn plan_default_expr(
...
+    let temp_storage = &RowArena::new();
+    for (mut param, ty) in params.into_iter().zip(&desc.param_types) {
+        transform_ast::transform_expr(scx, &mut param)?;
+        let expr = resolve_names_expr(&mut qcx, param)?;
+
+        let ecx = &ExprContext {
+            qcx: &qcx,
+            name: "EXECUTE",
+            scope: &scope,
+            // TODO(mjibson): Is this correct?
```
Comment (benesch) [inline src/sql/src/plan/query.rs:943]: Looks right to me since there's no relation in scope here!

File: src/sql/src/plan/query.rs [hunk 2]
```diff
@@ -913,6 +913,59 @@ pub fn plan_default_expr(
...
+            .lower_uncorrelated()?;
+        let evaled = ex.eval(&[], temp_storage)?;
+        datums.push(evaled);
+    }
+    Ok(Params {
+        datums,
+        types: desc
+            .param_types
+            .iter()
+            .map(|t| t.to_scalar_type_lossy())
```
Comment (benesch) [inline src/sql/src/plan/query.rs:964]: This looks sketchy though! (The call on L953 also looks sketchy.) Quite possibly it'll mishandle numeric scales or chars or something.

Going from a `pgrepr::Type` to a `ScalarType` is lossy, but going from a `ScalarType` to a `pgrepr::Type` is not. So I think instead you might want to `plan_expr` _without_ attempting to cast to anything. You probably want `.type_as_any()`. And _then_ you probably want to take the type of that expr, convert it to a `pgrepr::Type` and make sure it matches what you expect. Some very rough psuedocode that might be clearer than that verbiage:

```rust
let expr = plan_expr(ecx, &expr).type_as_any(ecx)?;
if pgrepr::Type::from(ecx.scalar_type(expr)) != ty {
    bail!("mismatched blah blah");
}
```

**Comment by benesch** [comment]
LGTM modulo the calls to `to_scalar_type_lossy`.

File: src/sql/src/plan/query.rs (hunk 2 above)
Comment (mjibson) [reply src/sql/src/plan/query.rs]: TIL! By the the hardest part of this PR was dealing with exprs and scalar/pgrepr types. Thanks for explaining the correct way.

**Pull request merged**